chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/Contexts.scala

```diff
@@ -0,0 +1,132 @@
+package io.scalaland.chimney.internal.compiletime
+
+import io.scalaland.chimney.{PartialTransformer, Patcher, Transformer}
+import io.scalaland.chimney.partial
+
+import scala.annotation.nowarn
+
+@nowarn("msg=The outer reference in this type test cannot be checked at run time.")
+private[compiletime] trait Contexts { this: Definitions & Configurations =>
+
+  /** Lazily evaluated log entry */
+  final protected class LogEntry(val nesting: Int, thunk: => String) {
+    lazy val message: String = thunk
+  }
+
+  sealed protected trait Context {
+
+    type Target
+    type Typeclass
+
+    def logs: Vector[LogEntry]
+    def appendLog(msg: => String): this.type
+  }
+
+  protected object Context {
+    sealed trait ForTransformer[From, To] extends Context {
+      val From: Type[From]
+      val To: Type[To]
+      val src: Expr[From]
+
+      val config: TransformerConfig[From, To]
+    }
+
+    final case class ForTotal[From, To](
+        From: Type[From],
+        To: Type[To],
+        src: Expr[From],
+        config: TransformerConfig[From, To],
+        logs: Vector[LogEntry],
+        logNesting: Int
+    ) extends ForTransformer[From, To] {
+
+      final type Target = To
+      final type Typeclass = Transformer[From, To]
+
+      override def appendLog(msg: => String): this.type =
+        copy(logs = logs :+ new LogEntry(nesting = logNesting, thunk = msg)).asInstanceOf[this.type]
+    }
+    object ForTotal {
+
+      def create[From: Type, To: Type](src: Expr[From], config: TransformerConfig[From, To]): ForTotal[From, To] =
+        ForTotal(
+          From = implicitly[Type[From]],
+          To = implicitly[Type[To]],
+          src = src,
+          config = config,
+          logs = Vector.empty,
+          logNesting = 0
+        )
+    }
+
+    final case class ForPartial[From, To](
+        From: Type[From],
+        To: Type[To],
+        src: Expr[From],
+        failFast: Expr[Boolean],
+        config: TransformerConfig[From, To],
+        logs: Vector[LogEntry],
+        logNesting: Int
+    ) extends ForTransformer[From, To] {
+
+      final type Target = partial.Result[To]
+      final type Typeclass = PartialTransformer[From, To]
+
+      override def appendLog(msg: => String): this.type =
+        copy(logs = logs :+ new LogEntry(nesting = logNesting, thunk = msg)).asInstanceOf[this.type]
+    }
+    object ForPartial {
+
+      def create[From: Type, To: Type](
+          src: Expr[From],
+          failFast: Expr[Boolean],
+          config: TransformerConfig[From, To]
+      ): ForPartial[From, To] = ForPartial(
+        From = implicitly[Type[From]],
+        To = implicitly[Type[To]],
+        src = src,
+        failFast = failFast,
+        config = config,
+        logs = Vector.empty,
+        logNesting = 0
+      )
+    }
+
+    final case class ForPatcher[T, Patch](
```

Comment:
Do you see any benefits from shared context base type for patchers and transformers?

My intuition is that the derivation code for both TCs will be quote independent (perhaps one calling other, but that would require building context from scratch anyway).

Reply:
Patchers are already pulling in Transformers so they would use potentially all of transformers functionality. They would surely share:
 * Types
 * Exprs
 * Results

Then we can theoretically turn `Definitions` into `TransformerDefinitions` and exclude `Patcher`s but it would inevitable require us to do:

```scala
trait TransformerDefinitions
  extends Types
  with Exprs
  with TransformerConfigurations
  with TransformerContexts
  with Results {
 ...
}

trait PatcherDefinitions
  extends TransformerDefinitions
  with PatcherConfigurations
  with PatcherContexts
```

and I am not sure there would be many benefits in such granularity.